### ISSUE TITLE
Bundled six fails to import when frozen

### ISSUE THREAD

**Issue by jdreaver** [vispy/vispy#1074]
When I freeze a simple vispy program, I can't import vispy because the import logic in `vispy.ext.six` fails when frozen. I have been trying to figure out how to fix this, but I couldn't in a reasonable amount of time so I will post my progress here.

Here is the script I am using to freeze the program (`freeze.py`):

```python
from cx_Freeze import setup, Executable

setup_kwargs = {
    'scripts': ['program.py'],
    'executables': [Executable("program.py")],
}

setup(**setup_kwargs)

```

Here is the one-line "program" that just imports vispy (`program.py`):

```python
import vispy
```

To freeze and run, just install `vispy` and `cx_Freeze` and type:

```sh
$ python freeze.py build_exe
$ ./build/exe.linux-x86_64-3.4/program 
```

(Your build folder is going to be different on a non-Linux machine or if you don't use Python 3.4).

This fails when vispy tries to do `imp.find_module('vispy.ext._bundled.six')`. I tried importing `six` like some of the other bundled packages if `hasattr(sys, "frozen")` is true in `vispy/ext/six.py`...:

```python
...

if hasattr(sys, "frozen"):
    try:
        from ._bundled.six import *  # noqa
    except ImportError:
        from six import *  # noqa
else:
    _import_six()

...
```

...but that didn't work. I got this error:

```
ImportError: No module named 'vispy.ext.six.moves'; 'vispy.ext.six' is not a package
```

I know @QuLogic did the bundling; do you have any ideas?

**Comment by almarklein** [comment]
Did you try adding `vispy.ext.six` to the list of cx_freeze includes? Althought that one is just a module that is imported in several places. I would, however, expect it to be necessary to add `vispy.ext._bundled.six` to the cx_freeze includes, as its imported dynamically.

**Comment by jdreaver** [comment]
I have tried both of those. I manually verified that the `vispy.ext.six` and `vispy.ext._bundled.six` modules were present in the zip file. I was wondering if @QuLogic had input on what is different between the bundled `six` and the other bundled modules in terms of importing functionality. I was under the impression that cx_Freeze uses `imp.find_module` itself, but for some reason in this context it fails.

I also tried just installing `six` separately and including it in the frozen build, but the `imp.find_module` call still fails. I will try to investigate this further in the next couple of weeks.

**Comment by Eric89GXL** [comment]
I think `six` might be different because we need a certain minimum version of `six` for compatibility.

**Comment by QuLogic** [comment]
There is no difference between the bundled `six` and the others, but you are having problems with the shim. You can try leaving out the bundled copy and only using a separate copy, but the shim might still give you trouble.

The shim was borrowed from Astropy and uses some crazy magic to get `six` working. I am not familiar with cx_Freeze and why it might be interfering with it. The "not a package" suggests there's a missing `__init__.py` but it should be there (unless cx_Freeze is missing it.) Perhaps it's because `imp` is deprecated in Python 3? Does it work on a Python 2 bundle?

**Comment by jdreaver** [comment]
I got an example application to work! Here is a summary of what I did:

* The only way I could get vispy to freeze was to undo the changes @QuLogic made to bundled libraries. More specifically, I removed the shims for `six` and `png` and moved the `six.py` and `png.py` files directly into `vispy/ext`.

* I had to manually add the glsl files to the freeze script, and add a modification to `vispy/glsl/__init__.py`:

```python
...
if hasattr(sys, "frozen"):
    path = op.join(op.dirname(sys.executable), "vispy", "glsl")
else:
    path = op.dirname(__file__) or '.'
...
```

* I did a similar thing for `vispy.io._data`. Here is the change to `vispy/io/datasets.py`:

```python
if hasattr(sys, "frozen"):
    DATA_DIR = op.join(op.dirname(sys.executable), 'vispy', 'io', '_data')
else:
    DATA_DIR = op.join(op.dirname(__file__), '_data')
```

* I had to specify that I wanted to include`vispy.app.backends`.

Here is my final `freeze.py` file:

```python
from cx_Freeze import setup, Executable

import os
import vispy.glsl
import vispy.io

glsl_folder = os.path.dirname(vispy.glsl.__file__)
io_folder = os.path.dirname(vispy.io.__file__)
io_data_folder = os.path.join(io_folder, "_data")
data_files = [
    (glsl_folder, os.path.join("vispy", "glsl")),
    (io_data_folder, os.path.join("vispy", "io", "_data")),
]

setup_kwargs = {
    'scripts': ['program.py'],
    'options': {
        'build_exe': {
            'include_files': data_files,
            'packages': [
                'vispy.app.backends',
            ],
        },
    },
    'executables': [Executable("program.py")],
}

setup(**setup_kwargs)
```

The only real problem I see in integrating these changes is the shims for the bundled packages. I'm afraid I would have to dive deep into cx_Freeze to figure out why they don't work, and I don't have that kind of time :smile:. I guess we have to decide if allowing dependencies to be unbundled is more important than allowing users to freeze code. It would be a shame if vispy couldn't be deployed in a frozen application.

Also, we could use `pkg_resources` to abstract away loading data files. Then, users wouldn't have to manually add the files to their freeze script, and we wouldn't need to add `if hasattr(sys, "frozen")` everywhere we load data.

**Comment by jdreaver** [comment]
Lastly, [here](https://github.com/jdreaver/vispy/commit/01dab6b689fd2fe01d7e69560d3205648cbd8715) is a link to the full commit I made for reference.

**Comment by Eric89GXL** [comment]
@QuLogic are the shims we have used by repos other than `astropy` (i.e., are they very common)? Is the purpose to make Debian/Fedora packaging easier?

**Comment by jdreaver** [comment]
I should also note that it seems like astropy [can't be frozen either](https://github.com/astropy/astropy/pull/960), so I don't know if they have successfully worked around the shim problems.

**Comment by opekar** [comment]
I used following dirty workaround with cx_Freeze

```python
import six
 ....
setup_kwargs = {
    'scripts': ['program.py'],
    'options': {
        'build_exe': {
                 include_files': [six.__file__, ... other included files]
...
```

The problem with this approach is that this includes the six.py file which is later compiled to .pyc (created with the first run of binary release). Not sure if this is any help , definitely not so clean.

**Comment by QuLogic** [comment]
The problem with Astropy is unrelated; it's much older than when they included the `six` shim.

If you print the exception [here](https://github.com/vispy/vispy/blob/master/vispy/ext/six.py#L42), you can see that the shim fails to import `vispy`. That's probably a bug in cx_Freeze where it overrides `importlib` but not `imp`. However, even if you switch to `importlib` (which can't be done unilaterally since Vispy support 2.6), you run into the same "not a package" error above.

I think this can be worked around, but `six` 1.4.1 uses a completely different method to create `six.moves` than later versions. Can we bump the requirement to something more recent?

**Comment by Eric89GXL** [comment]
It should be okay to bump the `six` requirement

**Comment by jdreaver** [comment]
Thanks for the info @QuLogic!

**Comment by jdreaver** [comment]
It looks like @QuLogic fixed the `six` error in #1136! 

I'm going to toy with using `pkg_resources` or some other package to embed the glsl and io data files better, so those will work automatically when frozen.
